In [1]:
#
import pandas as pd
pd.set_option('display.max_columns', None)

#
import numpy as np

#
import warnings

## 01. Exploración de datos.

In [2]:
df_flight_activity = pd.read_csv("../data/Customer Flight Activity.csv")
df_flight_activity.shape

(405624, 10)

In [3]:
df_loyalty = pd.read_csv("../data/Customer Loyalty History.csv")
df_loyalty.shape

(16737, 16)

### DF1: Flight Activity

In [4]:
df_flight_activity.head()

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
0,100018,2017,1,3,0,3,1521,152.0,0,0
1,100102,2017,1,10,4,14,2030,203.0,0,0
2,100140,2017,1,6,0,6,1200,120.0,0,0
3,100214,2017,1,0,0,0,0,0.0,0,0
4,100272,2017,1,0,0,0,0,0.0,0,0


In [5]:
df_flight_activity.tail()

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated,Points Redeemed,Dollar Cost Points Redeemed
405619,999902,2018,12,0,0,0,0,0.0,0,0
405620,999911,2018,12,0,0,0,0,0.0,0,0
405621,999940,2018,12,3,0,3,1233,123.0,0,0
405622,999982,2018,12,0,0,0,0,0.0,0,0
405623,999986,2018,12,0,0,0,0,0.0,0,0


*Observations on the DataFrame:*

* ⬜ Assess whether "Points Redeemed" and "Dollar Cost" are useful features.

In [6]:
df_flight_activity.info()

<class 'pandas.DataFrame'>
RangeIndex: 405624 entries, 0 to 405623
Data columns (total 10 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   Loyalty Number               405624 non-null  int64  
 1   Year                         405624 non-null  int64  
 2   Month                        405624 non-null  int64  
 3   Flights Booked               405624 non-null  int64  
 4   Flights with Companions      405624 non-null  int64  
 5   Total Flights                405624 non-null  int64  
 6   Distance                     405624 non-null  int64  
 7   Points Accumulated           405624 non-null  float64
 8   Points Redeemed              405624 non-null  int64  
 9   Dollar Cost Points Redeemed  405624 non-null  int64  
dtypes: float64(1), int64(9)
memory usage: 30.9 MB


*DataFrame Info Observations:*

- No null values detected in the DF.
- Column names follow a consistent style: they include spaces and use Title Case.
- "Year" and "Month" are stored as int instead of datetime: 
    * Decision: Maintained as int since the dataset lacks full date strings (missing days). Keeping them as integers ensures lighter performance and smoother DataFrame operations.
- Significant difference in total entries between DataFrames: he same customer can appear multiple times across different flights.
    * ⬜ Check for overall duplicate records in the DataFrame.
    * ⬜ Count the number of unique customers in the dataset.

In [7]:
df_flight_activity['Loyalty Number'].unique().shape

(16737,)

In [8]:
df_flight_activity.duplicated().sum()

np.int64(1864)

In [9]:
df_flight_activity.drop_duplicates(inplace=True)

In [10]:
df_flight_activity.duplicated().sum()

np.int64(0)

- ✅ Check for overall duplicate records in the DataFrame.
    * Update: Duplicates verified and dropped.

- ✅ Count the number of unique customers in the dataset.
    * Update: Validated; the unique customer count matches the secondary DataFrame.

In [12]:
df_flight_activity['Points Redeemed'].value_counts()

Points Redeemed
0      379579
447       103
523       100
443        98
516        97
        ...  
834         1
869         1
858         1
862         1
859         1
Name: count, Length: 587, dtype: int64

In [13]:
df_flight_activity['Dollar Cost Points Redeemed'].value_counts()

Dollar Cost Points Redeemed
0     379579
36      1037
38      1001
40       993
42       983
39       943
44       923
41       913
37       887
43       877
35       852
34       844
32       817
33       800
45       779
30       772
46       753
31       722
47       675
48       638
28       598
29       596
50       595
49       568
52       482
51       435
27       421
54       402
26       379
53       377
56       365
55       343
57       274
58       265
60       260
59       219
25       209
63       175
61       175
62       155
64       152
24       110
66       106
65        90
68        85
67        68
70        38
69        28
71         2
Name: count, dtype: int64

- ✅ Assess whether "Points Redeemed" and "Dollar Cost" are useful features.
    * Update: Evaluated both columns; although they contain valid non-null entries, they are non-informative for the scope of this analysis. Therefore, a decision was made to drop both features.

In [15]:
df_flight_activity.drop(columns=['Dollar Cost Points Redeemed','Points Redeemed'], inplace=True)

In [16]:
df_flight_activity.head()

,Loyalty Number,Year,Month,Flights Booked,Flights with Companions,Total Flights,Distance,Points Accumulated
0,100018,2017,1,3,0,3,1521,152.0
1,100102,2017,1,10,4,14,2030,203.0
2,100140,2017,1,6,0,6,1200,120.0
3,100214,2017,1,0,0,0,0,0.0
4,100272,2017,1,0,0,0,0,0.0


### DF2: Loyalty History

In [17]:
df_loyalty.head()

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Cancellation Year,Cancellation Month
0,480934,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN
1,549612,Canada,Alberta,Edmonton,T3G 6Y6,Male,College,NaN,Divorced,Star,3839.61,Standard,2016,3,NaN,NaN
2,429460,Canada,British Columbia,Vancouver,V6E 3D9,Male,College,NaN,Single,Star,3839.75,Standard,2014,7,2018.0,1.0
3,608370,Canada,Ontario,Toronto,P1W 1K4,Male,College,NaN,Single,Star,3839.75,Standard,2013,2,NaN,NaN
4,530508,Canada,Quebec,Hull,J8Y 3Z5,Male,Bachelor,103495.0,Married,Star,3842.79,Standard,2014,10,NaN,NaN


In [18]:
df_loyalty.info()

<class 'pandas.DataFrame'>
RangeIndex: 16737 entries, 0 to 16736
Data columns (total 16 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Loyalty Number      16737 non-null  int64  
 1   Country             16737 non-null  str    
 2   Province            16737 non-null  str    
 3   City                16737 non-null  str    
 4   Postal Code         16737 non-null  str    
 5   Gender              16737 non-null  str    
 6   Education           16737 non-null  str    
 7   Salary              12499 non-null  float64
 8   Marital Status      16737 non-null  str    
 9   Loyalty Card        16737 non-null  str    
 10  CLV                 16737 non-null  float64
 11  Enrollment Type     16737 non-null  str    
 12  Enrollment Year     16737 non-null  int64  
 13  Enrollment Month    16737 non-null  int64  
 14  Cancellation Year   2067 non-null   float64
 15  Cancellation Month  2067 non-null   float64
dtypes: float64(4), 

In [24]:
round(df_loyalty.isnull().sum()/df_loyalty.shape[0]*100, 2)

Loyalty Number         0.00
Country                0.00
Province               0.00
City                   0.00
Postal Code            0.00
Gender                 0.00
Education              0.00
Salary                25.32
Marital Status         0.00
Loyalty Card           0.00
CLV                    0.00
Enrollment Type        0.00
Enrollment Year        0.00
Enrollment Month       0.00
Cancellation Year     87.65
Cancellation Month    87.65
dtype: float64

*DataFrame Info Observations:*
 
- Column names follow a consistent style: they include spaces and use Title Case. Also consistent with the other DataFrame.
    * ⬜ Verify the number of unique customers in the dataset.
- The "Salary" column presents a significant missing data rate of 25%.
    * ⬜ Add to the null value handling pipeline.
- "Cancellation Year" and "Cancellation Month" exhibit a high missingness rate of 87.65%. This indicates they represent active customers, meaning null imputation is not required. Additionally, the columns are stored as float instead of int, because of the presence of null values.
    * ⬜ Assess whether these columns are useful features for the scope of the study.

In [26]:
df_loyalty['Loyalty Number'].value_counts().sum()

np.int64(16737)

In [27]:
df_loyalty.duplicated().sum()

np.int64(0)

* ✅ Verify the number of unique customers in the dataset:
    - Update: Confirmed; zero duplicate entries found.
* ✅ Assess whether these columns are useful features for the scope of the study:
    - Update: For the purpose of this study, the 'Cancellation Year' and 'Cancellation Month' columns are not required. However, a new feature will be created to preserve the customer's active status.

In [ ]:
# Created a new column indicating whether a customer is active or not by handling the null values.
df_loyalty['Customer Active'] = df_loyalty['Cancellation Year'].isnull()

In [30]:
df_loyalty.drop(columns=['Cancellation Year','Cancellation Month'], inplace=True)

In [31]:
df_loyalty.head()

,Loyalty Number,Country,Province,City,Postal Code,Gender,Education,Salary,Marital Status,Loyalty Card,CLV,Enrollment Type,Enrollment Year,Enrollment Month,Customer Active
0,480934,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,True
1,549612,Canada,Alberta,Edmonton,T3G 6Y6,Male,College,NaN,Divorced,Star,3839.61,Standard,2016,3,True
2,429460,Canada,British Columbia,Vancouver,V6E 3D9,Male,College,NaN,Single,Star,3839.75,Standard,2014,7,False
3,608370,Canada,Ontario,Toronto,P1W 1K4,Male,College,NaN,Single,Star,3839.75,Standard,2013,2,True
4,530508,Canada,Quebec,Hull,J8Y 3Z5,Male,Bachelor,103495.0,Married,Star,3842.79,Standard,2014,10,True
